src/openapi-generator.ts

```diff
@@ -353,22 +353,33 @@ export class OpenAPIGenerator {
     zodSchema: ZodSchema<any>
   ): SchemaObject | ReferenceObject {
     const innerSchema = this.unwrapChained(zodSchema);
-    const metadata = zodSchema._def.openapi
-      ? zodSchema._def.openapi
-      : innerSchema._def.openapi;
+    const metadata = zodSchema._def.openapi ?? innerSchema._def.openapi;
 
     const refId = metadata?.refId;
 
     if (refId && this.schemaRefs[refId]) {
-      const referenceObject = {
+      const schemaRef = this.schemaRefs[refId] as SchemaObject;
+      const referenceObject: ReferenceObject = {
         $ref: `#/components/schemas/${refId}`,
       };
 
-      const nullableMetadata = zodSchema.isNullable() ? { nullable: true } : {};
+      // New metadata from ZodSchema properties.
+      const newSchemaMetadata = omitBy(
+        this.toOpenAPISchema(innerSchema, zodSchema.isNullable()),
```

Comment:
In this same function we have a check if there is a provided `type` in the metadata and skip the generation logic in such case. I feel like the same should be applicable here. Basically in my opinion if a user opts to provide a `type` we should stop trying to infer properties since they might not be correct for his case.

For example `z.string().regex(/\d*/).openapi({ type: 'number' })` - the generated response should be `{ type: 'number' }` and not `{ type: 'number', pattern: '\d*' }`.